docs/protocols/xlang_object_graph_spec.md

```diff
@@ -0,0 +1,657 @@
++---------------+-------+-------+--------+-------+------------------------------------+
+| reserved bits |  oob  | xlang | endian | null  | unsigned int for meta start offset |
+```
+
+- null flag: 1 when object is null, 0 otherwise. If an object is null, other bits won't be set.
+- endian flag: 1 when data is encoded by little endian, 0 for big endian.
+- xlang flag: 1 when serialization uses xlang format, 0 when serialization uses Fury java format.
+- oob flag: 1 when passed `BufferCallback` is not null, 0 otherwise.
+
+If meta share mode is enabled, an uncompressed unsigned int is appended to indicate the start offset of metadata.
+
+## Reference Meta
+
+Reference tracking handles whether the object is null, and whether to track reference for the object by writing
+corresponding flags and maintaining internal state.
+
+Reference flags:
+
+| Flag                | Byte Value | Description                                                                                                                                             |
+|---------------------|------------|---------------------------------------------------------------------------------------------------------------------------------------------------------|
+| NULL FLAG           | `-3`       | This flag indicates the object is a null value. We don't use another byte to indicate REF, so that we can save one byte.                                |
+| REF FLAG            | `-2`       | This flag indicates the object is already serialized previously, and fury will write a ref id with unsigned varint format instead of serialize it again |
+| NOT_NULL VALUE FLAG | `-1`       | This flag indicates the object is a non-null value and fury doesn't track ref for this type of object.                                                  |
+| REF VALUE FLAG      | `0`        | This flag indicates the object is referencable and the first time to serialize.                                                                         |
+
+When reference tracking is disabled globally or for specific types, or for certain types within a particular
+context(e.g., a field of a type), only the `NULL` and `NOT_NULL VALUE` flags will be used for reference meta.
+
+For languages which doesn't support reference such as rust, reference tracking must be disabled for correct
+deserialization by fury rust implementation.
+
+For languages whose object values are not null by default:
+
+- In rust, Fury takes `Option:None` as a null value
+- In c++, Fury takes `std::nullopt` as a null value
+- In golang, Fury takes `null interface/pointer` as a null value
+
+If one want to deserialize in languages like `Java/Python/JavaScript`, he should mark the type with all fields
+not-null by default, or using schema-evolution mode to carry the not-null fields info in the data.
+
+## Type Meta
+
+For every type to be serialized, it must be registered with an optional ID first. The registered type will have a
```

Comment:
If the tag is replaced by type_id, does it mean that the order of registration needs to be consistent?

Reply:
Yes, users can specify an ID when registering, then the order doesn't have to be consistent. Otherwise, the order must be consistent.

Reply:
Specifing an ID is not that convenient compared to specify the tag, but it's faster and save space too.

Reply:
This do need more consideration, we must determine whether or not do this tradeoff